docs/style-guide.rst

```diff
@@ -0,0 +1,274 @@
+
+    * Cross-package imports **must not** request anything beyond the root namespace of the target package.
+    * Cross-package imports **may** be aliased where it aids readability.
+    * Cross-package imports **may** use ``from [module] import [package]`` syntax.
+
+    .. code-block:: python
+
+        # Good
+        from vyper.ast import fold
+        from vyper import ast as vy_ast
+
+        # Bad, do not import beyond the root namespace
+        from vyper.ast.annotation import annotate_python_ast
+
+Exceptions
+----------
+
+We use :ref:`custom exception classes <compiler-exceptions>` to indicate what has gone wrong during compilation.
+
+    * All raised exceptions **must** use an exception class that appropriately describes what has gone wrong. When none fits, or when using a single exception class for an overly broad range of errors, consider creating a new class.
+    * Builtin Python exceptions **must not** be raised intentionally. An unhandled builtin exception indicates a bug in the codebase.
+    * Use :func:`CompilerPanic<CompilerPanic>` for errors that are not caused by the user.
+
+Strings
+-------
+
+Strings substitutions **should** be performed via `formatted string literals <https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals>`_ rather than the ``str.format`` method or other techniques.
+
+Type Annotations
+----------------
+
+    * All publicly exposed classes and methods **should** include `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_ annotations for all arguments and return values.
+    * Type annotations **should** be included directly in the source. `Stub files <https://www.python.org/dev/peps/pep-0484/#stub-files>`_ **may** be used where there is a valid reason. Source files using stubs **must** still be annotated to aid readability.
+    * Internal methods **should** include type annotations.
+
+Tests
+=====
+
+We use the `pytest <https://docs.pytest.org/en/latest/>`_ framework for testing, and :ref:`eth-tester<testing-contracts-ethtester>` for our local development chain.
+
+Best Practices
+--------------
+
+    * ``pytest`` functionality **should not** be imported with ``from ...`` style syntax, particularly :func:`pytest.raises<pytest.raises>`. Importing the library itself aids readability.
+    * Tests **must not** be interdependent. We use ``xdist`` to execute tests in parallel. You **cannot** rely on which order tests will execute in, or that two tests will execute in the same process.
+    * Test cases **should** be designed with a minimalistic approach. Each test should verify a single behavior. A good test is one with few assertions, and where it is immediately obvious exactly what is being tested.
+    * Where logical, tests **should** be `parametrized <https://docs.pytest.org/en/latest/parametrize.html>`_ or use `property-based <https://hypothesis.works/>`_ testing.
+    * Tests **must not** involve mocking.
```

Comment:
```suggestion
    * Tests **must not** involve mocking, unless absolutely necessary.
```

Reply:
**should not** ?

Reply:
Maybe, I just don't think you can completely go without mocking sometimes, but definitely discourage it's use. Do we mock the CLI tests currently?

Reply:
No, CLI testing uses the `tmp_path` fixture to write test sources to disk prior to testing.

Reply:
Alright, forget I ever said this lol